Fix text bleeding on column layout

privatestefans opened the pull request:
# Fixed
- Fixed: Fix column mode page calculation that cause text bleeding (Fixes #4040)

## What was changed
- Use `outline` instead of `border` for visually debug elements to prevent any layout shifts
- Obtain the `offsetWidth` from the reader container rather than the reading section element to get the actual page width

## Preview after this fix
https://github.com/user-attachments/assets/b9a7f765-e203-4224-bf84-fb9d1bba039e

/claim #4040

privatestefans committed:
Use `outline` instead of `border` for visually debug elements to prevent layout shifts

privatestefans committed:
Fix `pageWidth` calculation when user switching from scroll mode and column(1,2) mode

privatestefans committed:
Fix go to next page after reaching the end of the page

privatestefans commented:
I can't see any issue with the vertical writing style for the Japanese version.
Even after adjusting the margin, it seems already correct. 🤔

<img width="1745" height="930" alt="brave_jQxNyG51PF" src="https://github.com/user-attachments/assets/bddea3cf-7586-4270-925a-2ef359ce86d6" />

majora2007 commented on UI/Web/src/app/book-reader/_components/book-reader/book-reader.component.scss:
```diff
@@ -256,6 +256,7 @@ $action-bar-height: 38px;
     overflow: hidden;
     word-break: break-word;
     overflow-wrap: break-word;
+    padding: 0 12px; // Add padding, so we can see the page border line on the left & right
```
The page border is only visible on debug mode (aka for local development), so this needs to be encapsulated.

privatestefans replied:
Ah, I see. If we don't need the page border to be visible, then we don't need this extra padding to prevent the text from being too close to the edge and slightly cut off. I will update it, thanks for the review 👍

majora2007 commented:
It seems you built off my latest attempt, which has received good feedback from testing. I'll pull this down and validate and get this tested out. 

The difference seems to mainly be switching from rounding to flooring and the padding logic, which needs to be in the .debug class, as that is not visible for the non-local build (switch the flag to validate).

privatestefans committed:
Remove extra padding and extra calculation to fix the page scroll

privatestefans commented:
I decided to remove the extra padding and calculation for now, as `column-rule: 2px dashed #666666;` is no longer add additional 2px for each column when debug flag is disabled.

### Updated preview
https://github.com/user-attachments/assets/ede1c434-4116-4625-b0fb-cb983770c6a0

majora2007 commented:
Did some testing on this and it's preforming slightly worse than my attempt for single column layout.

<img width="853" height="670" alt="image" src="https://github.com/user-attachments/assets/7206aab3-4c94-4978-a7f7-4fbbd93bb5cc" />

<img width="804" height="646" alt="image" src="https://github.com/user-attachments/assets/a6b13049-3495-4d6c-bd8a-4c28eec919af" />

This is only reproducible on this book (my attempt also had some slight page cutoff). 

page 14, single column, 15% margin, 1025x908.

https://wormhole.app/m9Oae3#hbOPG13nTuJsfGnPh0KZzA


I would say this is almost there.

privatestefans committed:
Reapply page styles on window resize and refresh column page scroll position

privatestefans commented:
Thanks for the sample file and some details.

After trying to reproduce the issue by changing the screen width, it seems the page margin did not update when the window resolution changed

I think we can fix it by reapplying page style with:
```js
this.applyPageStyles(this.pageStyles())
```

I can also reproduce bug in vertical layout during screen resizing. It occurs only when the window size changes and does not persist after a browser refresh. But I will investigate it further.

privatestefans committed:
Ensure `verticalBookContentWidth` updates properly when using `getVerticalPageWidth`

- Fix issue where `verticalBookContentWidth` returned stale values after window resize
- Convert `getVerticalPageWidth` to a computed signal so updates are properly detected

privatestefans committed:
Fix scroll snapping so it scrolls to the correct element

privatestefans commented:
Ref: https://github.com/Kareadita/Kavita/pull/4067/commits/87a64e9b3b60846dfcc97ca1d91e5e580aa40e4f
I have improved the `getFirstVisibleElementXPath` function to filter out empty elements and child elements that appear in the same entry list before returning the intended first visible element. I decided to improve this function because sometimes it returns an incorrect element (as you can see from the image below), and causing the page to snap to that element instead of the actual content or paragraph.

---
<img width="1299" height="569" alt="eTt462FoXp_scale" src="https://github.com/user-attachments/assets/83638f23-9856-476c-84c9-61113101367a" />

majora2007 commented:
Nice find, I do believe this will be an improvement. I've seen you're putting up a few commits. Just ping me when ready for me to retest.

privatestefans commented:
Sure thing!
I'm still trying to find another potential bug, but it hasn't shown up yet after a few adjustments I made.

privatestefans committed:
Some improvement, expand this commit message to see details

- Fix `Tap Pagination` button when using vertical writing style in scroll mode by relocating it to the `book-container` to prevent it from being stuck in the middle of the screen during scrolling
- Move `Tap Pagination` button to `book-container` to ensure visibility in vertical writing scroll mode
- Adjust button style to `top: 0px` to avoid affecting container size and which caused text bleeding
- Add warning to prevent race conditions when another scroll action still pending or delayed
- Ensure `PageHeightForPagination` used by the `Tap Pagination` button is updated whenever page layout or style changes occur
- Fix `addEmptyPageIfRequired` invocation and remove unnecessary empty pages after layout changes
- Fix page snapping and scrolling to specific positions

privatestefans committed:
Fix `Tap Pagination` button position

privatestefans commented:
Hi @majora2007, can you review this PR?

I’ve tested by switching between mobile and desktop views using the browser’s developer tools, and by changing the page layout from scroll mode to one- and two-column layout mode. So far, I haven’t noticed any remaining bugs during my testing.

I also made an attempt to improve the **Tap Pagination** button, especially for vertical writing style page:

<details>
<summary>Before the fix</summary>

https://github.com/user-attachments/assets/09bcc7aa-3427-47d5-b68b-c7b53262b3ff

</details>

<details>
<summary>After the fix</summary>

https://github.com/user-attachments/assets/441e08ea-ca89-4c27-b650-47612cfc48f4

</details>

majora2007 commented:
I like the new debug viewport mode, I'll pull it down today/tomorrow, if it passes my shakeout, I'll deploy this out to a dedicated environment and have a few testers validate.

Fesaa commented on UI/Web/src/app/book-reader/_components/book-reader/book-reader.component.ts:
```diff
@@ -1741,28 +1771,94 @@ export class BookReaderComponent implements OnInit, AfterViewInit, OnDestroy {
+        return !entries.some(item => element !== item && item.contains(element))
+
+        // Remove element that don't have any content
+        && (element.textContent.trim().length !== 0 || element.querySelectorAll('img, svg').length !== 0 || /^(img|svg)$/im.test(element.tagName));
```
`element.textContent.trim()` needs a nullcheck for `element.textContent`. (Throwns an error during compile)

Fesaa commented on UI/Web/src/app/book-reader/_components/book-reader/book-reader.component.ts:
```diff
@@ -1485,6 +1510,12 @@ export class BookReaderComponent implements OnInit, AfterViewInit, OnDestroy {
+    const bookContentElem = this.bookContentElemRef.nativeElement;
+
+    // Remove the empty gap first (if exist) to recalculate if we need the empty gap
+    const oldEmptyGap = bookContentElem.querySelector('.kavita-empty-gap');
```
I'm seeing some weird behaviour when paging backwards where the empty page is removed, then added again. But since I'm on the last page, it "scrolls" back, and the empty page causes you to need to page over the same text twice.

https://github.com/user-attachments/assets/60af97c7-2e3b-4bf4-8ef6-0d72b9ceb333

I'm not sure how we would best solve this. Do the calcs by subtracting the width of the empty page if present, and then remove/rescale as needed?

privatestefans replied:
Ah, I will figure out the best way to fix that. I think doing some calcs can solve it.
Thanks for the feedback, by the way 👍

Fesaa commented:
Heya,

Left some comments on issues we've found. Overall, looks really promising! If you could have a look at them (only one really, the other is a small change) and merge in develop, so your branch is up-to-date. After which we'll deploy and have other testers test as well.

During testing, we've also noticed the following issues, which we'll have a look at as well.
- Bookmark indicators not being visible (regressing somewhere)
- Invalid XPath exception when loading a page with only an image

Thanks!

privatestefans commented:
Sure, I will look into it and I’m still figuring out the empty page issue at the moment.

It seems like the bookmark indicator has shifted to another page.
I'm not sure what’s causing this yet, but I might be able to check it out.

Thanks @Fesaa!

privatestefans committed:
Fix empty page gap handling and bookmark position

- Also add null check to `element.textContent`

privatestefans commented:
About the bookmark indicator.

I noticed that some image elements have their `position` overridden by a non-existent CSS variable. As a result, the element `position` default to 'static', which causes all bookmark attached the image elements with `.kavita-scale-width-container` become incorrectly positioned.

<img width="1411" height="659" alt="q9q8G4Z2sh" src="https://github.com/user-attachments/assets/8beda3d1-f89b-4eef-b927-baac193bb48b" />

So I set `--book-reader-content-position` to 'relative' as the default for all image elements, except for the single-page element, since I think there’s no issue with that one.

Fesaa commented:
Heya,

To keep you in the loop, the empty page issue is resolved. So your code has been deployed to have testers test it. :tada: 
I've merged develop in before deploying, could you also do this for your branch. Thanks

Bookmarks / images are not, I'm seeing some rather odd behaviour. But I do not believe it's caused by your changes. I will be looking into those. 

Thanks again!

privatestefans committed:
Merge branch 'Kareadita:develop' into develop

privatestefans commented:
Thanks for letting me know. I've synced my branch with develop.

Also, I added a default value for `--book-reader-content-position` in my last commit. I haven't tested it with any book other than the one I have, though.

But feel free to ask me to make changes if needed 👍

majora2007 commented:
Doing some extended testing on this and facing a few edge cases. I tested a few books, only this ravens of the inner palace (image heavy) had issues. I can work on the bookmarking positioning, but the image cutoff and next page seem like regressions. 

Overall, from a text perspective, this seems to fix the epub bleed. If we can cleanup these small regressions, we can close this out. 

## Ravens of the Inner Palace

#### Next page broken
<img width="1909" height="879" alt="image" src="https://github.com/user-attachments/assets/a5722331-9292-4587-9825-29f5e6610e77" />
> I cannot move to the next page. `currentVirtualPage` is always 1 and the `totalVirtualPages` is 2.
> Page is also laying out incorrectly. It should be centered with 1 column, not right aligned as if in 2 column layout. (works correctly in 2 column and scroll)


#### Image cutoff
<img width="1597" height="901" alt="image" src="https://github.com/user-attachments/assets/9ea0c097-f8a0-4db2-8020-b50b0ab44245" />

<img width="1791" height="264" alt="image" src="https://github.com/user-attachments/assets/17994257-3b43-4076-98f1-e875a5006e26" />

> Images are being cutoff, this seems to be due to `--book-reader-content-max-height` on the surrounding image container being too high (532px on mine, bringing it down to near `55px - 38px` fixes the layout. 


#### Bookmark icon cutoff
<img width="87" height="72" alt="image" src="https://github.com/user-attachments/assets/ffda32bf-6ec3-4a97-8037-14bc4aa4b199" />

> The bookmark icon calculation needs some slight tweaking. It seems to be a few pixels off which is causing some cutoff. 

<img width="1209" height="650" alt="image" src="https://github.com/user-attachments/assets/d1e13aae-2b89-41fb-ba63-b4ba2d399f9b" />

> 2 column layout also has some weird cutoff issue that is fixed with even more offset. (I fixed with a -10 on top/left)

privatestefans committed:
Revert content position default value and improve element filter

privatestefans commented:
Whoops, looks like using `relative` as the default value for `book-reader-content-position` for fixing the bookmark is not a good idea for that book. I've reverted it and also added a small improvement for filtering element.

Can you recheck it @majora2007?

majora2007 commented:
That fixed all my issues. It's looking great, I'll do another shakeout with some more books to validate, but we might be done with this issue.

majora2007 commented:
Great news, everything checked out. We had found some small strangeness with japanese books (scroll position shifting between layout mode). This is something we are willing to accept as an edge case, likely just better `firstVisibleItem` code in the future.

Please submit an expense against OC for 200$ USD.

privatestefans commented:
I’m glad to hear the fix worked well! Thank you for reviewing it carefully. I agree `getFirstVisibleElement` could be refined. I also spent some time looking into that, but I was careful not to over-filter since it might not work for some books.

I've just submitted the expense through OpenCollective, thank you!

majora2007 commented:
Agreed, it's a balancing act when books have the ability to put any styles and html in them. I just approved the OC, thank you again.
